tests/test_tokens_prices.py

```diff
@@ -0,0 +1,124 @@
+from coingecko_service import *
+from conftest import SERVICE_MAP
+from conftest import SERVICE_COINGECKO
+from constants import *
+
+class TestTokensPrices:
+
+
+
+    # NUM_TOKENS_DEFAULT low liquidity tokens
+    @pytest.mark.parametrize("token",setup.choose_tokens_liq_range(NUM_TOKENS_DEFAULT, MIN_LIQ_FILTER_DEFAULT, MAX_VAL_LOW_LIQ_FILTER_DEFAULT))
+    def test_low_liq_token_prices(self, environment_url, token):
+        self.run_coingecko_comparison_test(environment_url, token, HIGH_PRICE_DIFF)
+
+    # NUM_TOKENS_DEFAULT low volume tokens
+    @pytest.mark.parametrize("token",setup.choose_tokens_volume_range(NUM_TOKENS_DEFAULT, MIN_VOL_FILTER_DEFAULT, MAX_VAL_LOW_VOL_FILTER_DEFAULT))
+    def test_low_volume_token_prices(self, environment_url, token):
+        self.run_coingecko_comparison_test(environment_url, token, HIGH_PRICE_DIFF)
+
+    # NUM_TOKENS_DEFAULT mid volume tokens
+    @pytest.mark.parametrize("token",setup.choose_tokens_volume_range(NUM_TOKENS_DEFAULT, MIN_VOL_FILTER_DEFAULT, MAX_VAL_MID_VOL_FILTER_DEFAULT))
+    def test_mid_volume_token_prices(self, environment_url, token):
+        self.run_coingecko_comparison_test(environment_url, token, MID_PRICE_DIFF)
+
+    # NUM_TOKENS_DEFAULT top by-volume tokens
+    @pytest.mark.parametrize("token", setup.choose_tokens_volume_range(NUM_TOKENS_DEFAULT))
+    def test_top_volume_token_prices(self, environment_url, token):
+        self.run_coingecko_comparison_test(environment_url, token, LOW_PRICE_DIFF)
+
+    # NUM_TOKENS_DEFAULT top by-volume tokens in a batch request, in which multiple tokens
+    # are requested in a single request to /tokens/prices
+    def test_top_volume_token_prices_in_batch(self, environment_url):
+        tokens = setup.choose_tokens_volume_range(NUM_TOKENS_DEFAULT)
+        date_format = '%Y-%m-%d %H:%M:%S'
+        sqs_service = SERVICE_MAP[environment_url]
+        # Assert the latency of the sqs pricing request is within the threshold
+        measure_latency = lambda: sqs_service.get_tokens_prices(tokens)
+        latency = timeit.timeit(measure_latency, number=1)
+        print(f"{datetime.fromtimestamp(time.time()).strftime(date_format)}: sqs pricing response time = {latency}")
+        assert latency < RT_THRESHOLD, f"SQS pricing request response time {latency} exceeds {RT_THRESHOLD} second"
+
+        # Assert sqs price is available for the token
+        sqs_price_json = sqs_service.get_tokens_prices(tokens)
+        for token in tokens:
+            sqs_price_str = sqs_price_json.get(token, {}).get(USDC, None)
+            assert sqs_price_str is not None, f"{token} SQS price is none"
+            sqs_price = float(sqs_price_str)
+            assert sqs_price > 0, f"{token} SQS price is zero"
+
+    # Test every valid listed token if it is supported by the /tokens/prices endpoint
+    # Count the number of unsupported tokens and assert it is below the threshold
+    def test_unsupported_token_count(self, environment_url):
```

Comment:
Could we parameterize this test case similar to others so that we can run them in parallel while at the same time having the assertion about the unsupported count?

We could have an atomic integer in the global space that each test case would update and then a separate test fixture that would assert on its value relative to the threshold.

As of right now, we would be running each in sequence, increasing the CI time

Reply:
It looks like pytest-xdist runs the tests in different processes but not threads and so we need to do some inter-process communication. I can think of using FileLock to do so. Do you have some better idea?

Reply:
Got it, interesting.

From my investigations, most suggestions propose similar solutions.

We could then simply spawn Python threads within a single test case. While not parallel, we would still benefit from concurrency of not having to wait for blocking network requests, improving CI times.

Reply:
Issues resolved. Please check out the latest commit. Thanks!